### PR TITLE
Add view port effect for post processing

### PR DESCRIPTION
**Pull request by levinli303**
When we need to apply a viewport effect, we first render to a texture, then render the texture with effect on screen. This will be needed for fisheye warping.

Configure in celestia.cfg, two effects available:
1. Passthrough effect, just draw the same thing from the texture to the screen. `ViewportEffect "passthrough"`
2. Warp mesh effect, warping according to the warp mesh file provided. `ViewportEffect "warpmesh" WarpMeshFile "some warp mesh file name in the warp folder"`

The warp mesh effect and file format is from http://paulbourke.net/dataformats/meshwarp/ which is basically a mapping between vertex coordinates and texture coordinates.

### CODE REVIEW

**Comment by levinli303** [comment]
Need to be tested, so far it works ok on Mac and iOS.  Used [warp.zip](https://github.com/CelestiaProject/Celestia/files/5331887/warp.zip) for testing the warpmesh.

**Comment by 375gnu** [comment]
i would like not to have GL calls in any celestia/* files, could you move them to celengine/ introducing new helper classes if necessary. another problem is oldfbo in FBO class, this violates encapsulation rules.

**Comment by levinli303** [comment]
@375gnu  maybe we can just move the glClear(Color) to inside ViewportEffect? The others are not direct GL calls, except for generating FBOs which I can't seem to find a trivial way to move to celengine since it is tied to view.

**Comment by 375gnu** [comment]
> maybe we can just move the glClear(Color) to inside ViewportEffect?

yes, this sounds reasonable

**Comment by levinli303** [comment]
> > maybe we can just move the glClear(Color) to inside ViewportEffect?
> 
> yes, this sounds reasonable

this PR should be ready now.

**Comment by 375gnu** [comment]
@levinli303 i'm still isn't happy with FBO unbinding. Generally we need to know if there is any FBO bound only in `renderGeometry_GLSL` . We can save currently bound FBO id before calling `renderGeometryShadow_GLSL` and restore it after. In this case `FramebufferObject` will remain the same.

**Comment by levinli303** [comment]
I think this is how it is done here. I added this here because OpenGLES on iOS's default FBO to render to is not 0.

File: src/celengine/mapmanager.cpp [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+    }
+    // Texture coordinate is [0, 1], normalize to [-1, 1]
+    *u = (*u) * 2 - 1;
+    *v = (*v) * 2 - 1;
+    return true;
+}
+
+static WarpMeshManager* warpMeshManager = nullptr;
+
+extern WarpMeshManager* GetWarpMeshManager()
```
Comment (375gnu) [inline src/celengine/mapmanager.cpp:106]: please remove `extern` and move `static WarpMeshManager* warpMeshManager = nullptr;` into the function body

File: src/celengine/mapmanager.cpp [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        warpMeshManager = new WarpMeshManager("warp");
+    return warpMeshManager;
+}
+
+static string resolveWildcard(const string& filename)
+{
+    string base(filename, 0, filename.length() - 1);
+
+    string mapfile = base + "map";
+    {
```
Comment (375gnu) [inline src/celengine/mapmanager.cpp:118]: no need for braces here

File: src/celengine/mapmanager.cpp [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        ifstream in(mapfile);
+        if (in.good())
+            return mapfile;
+    }
+
+    return "";
```
Comment (375gnu) [inline src/celengine/mapmanager.cpp:124]: please use `return {}` here

File: src/celengine/mapmanager.cpp [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    return filename;
+}
+
+
+WarpMesh* WarpMeshInfo::load(const fs::path& name)
+{
+#define MESHTYPE_RECT 2
+    ifstream f(name.c_str());
```
Comment (375gnu) [inline src/celengine/mapmanager.cpp:150]: please change to `name.string()`

File: src/celengine/mapmanager.h [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+#include <string>
+#include <map>
+#include <functional>
+#include <celutil/resmanager.h>
+
+// File format for data used to warp an image, for
+// detail, see http://paulbourke.net/dataformats/meshwarp/
+class WarpMesh
+{
+public:
```
Comment (375gnu) [inline src/celengine/mapmanager.h:23]: please add a space before `public` & `private`

File: src/celengine/mapmanager.h [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+};
+
+inline bool operator<(const WarpMeshInfo& wi0, const WarpMeshInfo& wi1)
+{
+    return wi0.source < wi1.source;
+}
+
+typedef ResourceManager<WarpMeshInfo> WarpMeshManager;
+
+extern WarpMeshManager* GetWarpMeshManager();
```
Comment (375gnu) [inline src/celengine/mapmanager.h:58]: no need for `extern`

File: src/celengine/viewporteffect.cpp [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        return false;
+
+    glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
+    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+    return true;
+}
+
+bool ViewportEffect::distortXY(float &x, float &y) { return true; };
+
+VOViewportEffect::VOViewportEffect(GLenum bufferType, GLsizeiptr bufferSize, GLenum streamType) : vo(bufferType, bufferSize, streamType)
```
Comment (375gnu) [inline src/celengine/viewporteffect.cpp:35]: please move `vo(...)` to the next line

File: src/celengine/viewporteffect.cpp [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    prog->use();
+    configureShader(prog, width, height);
+    glBindTexture(GL_TEXTURE_2D, fbo->colorTexture());
+    draw(vo);
+    glBindTexture(GL_TEXTURE_2D, 0);
+    vo.unbind();
+    return true;
+}
+
+PassthroughViewportEffect::PassthroughViewportEffect() : VOViewportEffect(GL_ARRAY_BUFFER, 0, GL_STATIC_DRAW)
```
Comment (375gnu) [inline src/celengine/viewporteffect.cpp:58]: please move `VOViewportEffect(...)` to  the next line

File: src/celengine/viewporteffect.cpp [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+{
+    prog->samplerParam("tex") = 0;
+}
+
+void PassthroughViewportEffect::draw(celgl::VertexObject& vo)
+{
+    vo.draw(GL_TRIANGLES, 6);
+}
+
+WarpMeshViewportEffect::WarpMeshViewportEffect(WarpMesh *mesh) : VOViewportEffect(GL_ARRAY_BUFFER, 0, GL_STATIC_DRAW), mesh(mesh) {}
```
Comment (375gnu) [inline src/celengine/viewporteffect.cpp:103]: as above

File: src/celengine/mapmanager.h [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+// mapmanager.h
+//
+// Copyright © 2020 Celestia Development Team. All rights reserved.
+//
+// This program is free software; you can redistribute it and/or
+// modify it under the terms of the GNU General Public License
+// as published by the Free Software Foundation; either version 2
+// of the License, or (at your option) any later version.
+
+#ifndef _MAPMANAGER_H_
```
Comment (375gnu) [inline src/celengine/mapmanager.h:11]: please use `#pragma once`

File: src/celengine/viewporteffect.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+};
+
+class VOViewportEffect : public ViewportEffect
+{
+public:
+    VOViewportEffect(GLenum bufferType, GLsizeiptr bufferSize, GLenum streamType);
+    ~VOViewportEffect() override = default;
+
+    bool render(Renderer*, FramebufferObject*, int width, int height) override;
+    virtual std::string getShaderName() = 0;
```
Comment (375gnu) [inline src/celengine/viewporteffect.h:43]: this method should be `const`

File: src/celengine/viewporteffect.cpp [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+{
+    if (!fbo->unbind(oldFboId))
+        return false;
+
+    glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
+    glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+    return true;
+}
+
+bool ViewportEffect::distortXY(float &x, float &y) { return true; };
```
Comment (375gnu) [inline src/celengine/viewporteffect.cpp:33]: please make multiline

File: src/celengine/viewporteffect.h [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+    virtual bool preprocess(Renderer*, FramebufferObject*);
+    virtual bool prerender(Renderer*, FramebufferObject*);
+    virtual bool render(Renderer*, FramebufferObject*, int width, int height) = 0;
+    virtual bool distortXY(float& x, float& y);
+
+private:
+    GLint oldFboId;
+};
+
+class VOViewportEffect : public ViewportEffect
```
Comment (375gnu) [inline src/celengine/viewporteffect.h:36]: do we really need yet another virtual subclass?

**Review by 375gnu — CHANGES_REQUESTED** [review]
please add a description about warp algorithm used, at least a link to documentation

i'm still unhappy with FBO usage in CelestiaCore and FBO switching in unbind(), but at least for now we can live with this and will refactor this code later

**Review by 375gnu — APPROVED** [review]
let's roll

**Pull request merged**